<a href="https://colab.research.google.com/github/OhSujin/HelloSujinWorld/blob/master/n113a_data_manipulation_ipynb%EC%9D%98_%EC%82%AC%EB%B3%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src='https://user-images.githubusercontent.com/6457691/89746244-a6e2a200-daf3-11ea-9314-7863d544d73c.png' width = '200' align = 'right'>

## *DATA SCIENCE / SECTION 1 / SPRINT 1 / NOTE 3*

# 📝 Assignment

---

## 1. Concat 

세션시간에 다뤘던 6개의 데이터를 1개로 합치세요.

데이터를 불러오기 위하여 아래의 코드를 참조하세요.

합쳐진 데이터의 이름은 `df`입니다.

```python
import pandas as pd

urlhead = 'https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/stocks/'
df1 = pd.read_csv(urlhead + '000080.csv') 
# '000890.csv'
# '005300.csv'
# '027740.csv'
# '035810.csv'
# '136480.csv'
```



단, 모든 feature, 모든 데이터에 대해서가 아닌

- 매출액
- 자본총계
- EPS

- 제일 최근 분기의 데이터

에 대해서만 하면 됩니다.

||매출액|자본총계|EPS(원)|종목명|
|:-:|:-:|:-:|:-:|:-:|
|2020/09|6,243|10,897|453|하이트진로|
|...|...|...|...|...|
|2020/09|2,613|2,695|62|하림|

(첫부분은 값이 달라도 되며, 결과 shape는 6,4입니다.)

힌트 : 
- 데이터를 transpose
- column의 이름을 변경 및 설정
- 데이터의 일부 feature 선택
- `종목명`이라는 feature 지정
- 이후 `concat`으로 합치기
- 합친 이후의 index를 종목명으로 설정 및 drop

이라는 단계를 거치면 될겁니다.

In [18]:
def mydf(myurl):
    # 데이터 transpose
    df = pd.read_csv(urlhead + myurl).transpose()

    # index 처리
    new_header = df.iloc[0] 
    df = df[1:] 
    df.columns = new_header 

    # 데이터의 일부 feature 선택 => 제일 최근 분기의 데이터
    return df[-1:]

df = pd.concat([mydf('000080.csv'), mydf('000890.csv'), mydf('005300.csv'), mydf('027740.csv'), mydf('035810.csv'), mydf('136480.csv')])
df = df.reset_index()
df = df.drop(df.columns[0], axis=1)
    
df

Unnamed: 0    매출액 영업이익 영업이익(발표기준)  ... 현금배당수익률 현금배당성향(%)   발행주식수(보통주)
0           6,243  644        644  ...     NaN       NaN   70,133,611
1             205    6          6  ...     NaN       NaN  127,449,373
2           5,980  293        293  ...     NaN      0.00    7,993,460
3             521  -89        -89  ...     NaN      0.00  168,807,631
4           4,108  246        246  ...     NaN      0.00   55,640,623
5           2,613    1          1  ...     NaN       NaN  106,209,702

[6 rows x 33 columns]

In [19]:
df2 = pd.DataFrame(
    {'종목' : ['000080', '000890', '005300', '027740', '035810', '136480'],
     '종목명': ['하이트진로', '보해양조', '롯데칠성', '마니커', '이지홀딩스', '하림',], 
     '테마' : ['주류', '주류', '주류', '육계', '육계', '육계']}
)
df2

종목    종목명  테마
0  000080  하이트진로  주류
1  000890   보해양조  주류
2  005300   롯데칠성  주류
3  027740    마니커  육계
4  035810  이지홀딩스  육계
5  136480     하림  육계

In [20]:
df = pd.concat([df2, df], axis = 1)

In [21]:
df = df[['매출액','자본총계', 'EPS(원)','종목명']]
df

매출액    자본총계 EPS(원)    종목명
0  6,243  10,897    453  하이트진로
1    205     816      3   보해양조
2  5,980  12,096  1,682   롯데칠성
3    521     624    -61    마니커
4  4,108   7,534  1,168  이지홀딩스
5  2,613   2,695     62     하림

In [22]:
df.shape

(6, 4)

## 2. Reshape 

방금 만든 데이터를 아래 예시와 같은 **tidy** 형태로 만드세요.

이때 tidy 데이터의 이름은 `df_tidy` 입니다.

|종목명|Feature|Value|
|:-:|:-:|:-:|
|하이트진로|매출액|6243|
|보해양조|매출액|205|
|롯데칠성|매출액|5980|
|...|...|...|
|마니커|EPS(원)|-61|
|이지홀딩스|EPS(원)|1168|
|하림|EPS(원)|62|

이 후 아래의 코드를 정상적으로 실행했을때 나오는 box의 수를 `res`에 저장하세요.

```python
sns.boxplot(data = df_tidy, x = '종목명', y = 'value');
```




In [72]:
df_tidy = df.reset_index()
df_tidy = df_tidy.melt(id_vars= '종목명', value_vars= ['매출액','자본총계','EPS(원)'])
df_tidy

종목명 variable   value
0   하이트진로      매출액   6,243
1    보해양조      매출액     205
2    롯데칠성      매출액   5,980
3     마니커      매출액     521
4   이지홀딩스      매출액   4,108
5      하림      매출액   2,613
6   하이트진로     자본총계  10,897
7    보해양조     자본총계     816
8    롯데칠성     자본총계  12,096
9     마니커     자본총계     624
10  이지홀딩스     자본총계   7,534
11     하림     자본총계   2,695
12  하이트진로   EPS(원)     453
13   보해양조   EPS(원)       3
14   롯데칠성   EPS(원)   1,682
15    마니커   EPS(원)     -61
16  이지홀딩스   EPS(원)   1,168
17     하림   EPS(원)      62

In [74]:
# columns 이름 바꿔주기
df_tidy = df_tidy.rename(
    columns = {
        #'종목명': '종목명', 
        'variable': 'feature', 
        #'value': 'value'
        }
)
df_tidy

종목명 feature   value
0   하이트진로     매출액   6,243
1    보해양조     매출액     205
2    롯데칠성     매출액   5,980
3     마니커     매출액     521
4   이지홀딩스     매출액   4,108
5      하림     매출액   2,613
6   하이트진로    자본총계  10,897
7    보해양조    자본총계     816
8    롯데칠성    자본총계  12,096
9     마니커    자본총계     624
10  이지홀딩스    자본총계   7,534
11     하림    자본총계   2,695
12  하이트진로  EPS(원)     453
13   보해양조  EPS(원)       3
14   롯데칠성  EPS(원)   1,682
15    마니커  EPS(원)     -61
16  이지홀딩스  EPS(원)   1,168
17     하림  EPS(원)      62

In [78]:
# 박스의 수 res에 저장하기
res =6

## 3. Merge & Join +

pandas merge의 options중 how 에는 여러가지 옵션들이 있습니다.

공식 레퍼런스를 검색, 참조하여

아래 코드를 바탕으로 

다음과 같은 결과물을 만드는 코드를 작성하세요

변수의 이름은 `df2` 입니다.

|key1|key2|A|B|C|D|
|:-:|:-:|:-:|:-:|:-:|:-:|
|K0|K0|A0|B0|C0|D0|
|K0|K1|A1|B1|NaN|NaN|
|K1|K0|A2|B2|C1|D1|
|K1|K0|A2|B2|C2|D2|
|K2|K1|A3|B3|NaN|NaN|





In [52]:
import pandas as pd

left = pd.DataFrame({'key1': ['K0', 'K0', 'K1', 'K2'],
                     'key2': ['K0', 'K1', 'K0', 'K1'],
                     'A': ['A0', 'A1', 'A2', 'A3'],
                     'B': ['B0', 'B1', 'B2', 'B3']})

right = pd.DataFrame({'key1': ['K0', 'K1', 'K1', 'K2'],
                      'key2': ['K0', 'K0', 'K0', 'K0'],
                      'C': ['C0', 'C1', 'C2', 'C3'],
                      'D': ['D0', 'D1', 'D2', 'D3']})

df2 = left.merge(right, how = 'left')
df2

key1 key2   A   B    C    D
0   K0   K0  A0  B0   C0   D0
1   K0   K1  A1  B1  NaN  NaN
2   K1   K0  A2  B2   C1   D1
3   K1   K0  A2  B2   C2   D2
4   K2   K1  A3  B3  NaN  NaN

## Grading

이전 과제를 완료했다면, 아래의 코드를 실행하세요. 

In [80]:
import numpy as np

def grade(df, df2, df_tidy, res):
  score = 0
  score += (df['매출액'][2] == 5980)
  score += (df['자본총계'][3] == 1897)   
  score += (df2.equals(pd.DataFrame({
    'key1' : ['K0', 'K0', 'K1', 'K1', 'K2'],
    'key2' : ['K0', 'K1', 'K0', 'K0', 'K1'],
    'A' : ['A0', 'A1', 'A2', 'A2', 'A3'],
    'B' : ['B0', 'B1', 'B2', 'B2', 'B3'],
    'C' : ['C0', np.NaN, 'C1', 'C2', np.NaN],
    'D' : ['D0', np.NaN, 'D1', 'D2', np.NaN],
  }))) 
  score += (res == 6)
  score += (round(df_tidy['value'].mean(), 3) == 7090.444)
  return(score)

In [81]:
grade(df, df2, df_tidy, res)

TypeError: ignored

## 🔥 도전과제

아래 링크들을 참조하여, 
특정 테마별로 포함되어 있는 종목의 데이터를 직접 csv형태로 수집하고 (아래 적힌 3개씩)

<https://finance.naver.com/sise/sise_group_detail.nhn?type=theme&no=408>

<https://finance.naver.com/sise/sise_group_detail.nhn?type=theme&no=36>

- 최종 분기에 해당하는 데이터

- 테마별 평균 데이터를 결과물로 만드는 코드를 작성하세요.

단, 모든 feature에 대해서가 아닌

- 매출액
- 자본총계
- EPS

에 대해서만 하면 됩니다.

|종목명|매출액|자본총계|EPS|
|:-:|:-:|:-:|:-:|
|대한해운|...|...|...|...|
|팬오션|...|...|...|...|
|KSS해운|...|...|...|...|
|세미콘라이트|...|...|...|...|
|오성첨단소재|...|...|...|...|
|에이아이비트|...|...|...|...|

---

|테마|매출액|자본총계|EPS|
|:-:|:-:|:-:|:-:|
|해운|...|...|...|
|마리화나(대마)|...|...|...|

이후 아래의 예시 처럼 seaborn을 사용하여 결과를 barplot으로 시각화하세요.

단, 한글이 깨지는 에러를 여러분이 직접 고쳐야합니다.

In [ ]:
import pandas as pd
import seaborn as sns

dat = pd.DataFrame({
    'name' : ['해운', '대마'],
    'value' : [10000,5000]
})

In [ ]:
sns.barplot(x = 'name', y = 'value', data = dat);